### PR TITLE
NixOS EC2 AMI: Support IMDSv2

### PR DESCRIPTION
**Pull request by grahamc**
###### Motivation for this change

AWS's metadata service has two versions. Version 1 allowed plain HTTP requests to get metadata. However, this was frequently abused when a user could trick an AWS-hosted server in to proxying requests to the metadata service. Since the metadata service is frequently used to generate AWS access keys, this is pretty gnarly. Version two is identical except it requires the caller to request a token and provide it on each request.

Today, starting a NixOS AMI in EC2 where the metadata service is configured to only allow v2 requests fails: the user's SSH key is to placed, and configuration provided by the user-data is not applied. The server is useless. This patch addresses that.

OpenStack "borrowed" this metadata fetcher when it happened to be the same API, but this new version is different enough, imo, to warrant undoing this borrow.

~This PR is not yet tested in an actual deployment, and I'll mark it as ready once I have. It should be backported to stable and new AMIs should be built.~

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by andir — COMMENTED** [review]
Nice! Thank you for digging into this :+1: 

When you are done can you ensure that the commit messages have an equal amount of information as the PR description has right now? This will make it much easier for someone to debug this in the future as most editors support showing commit messages inline.

**Comment by grahamc** [comment]
Done, @andir. I've tested this with real AWS machines and it works correctly, fixing the issue preventing our AMI from working with IMDSv2.

**Review by andir — COMMENTED** [review]
I didn't run this but the solution really looks straight forward. Given that this is for AWS instances and not for my kettle's IoT interface the additional 30M do not seem too bad. If, in fact, this becomes an issue we could still use a slimmer version of curl or revisit our options here.

There are a few comments that I'd like to see clarified before I can "approve" this.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:21]: Yeah. It could easily succeed within a second or so. Blocking early boot for 30s at a time seems intense.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:38]: Not sure we should abort. Maybe somebody has made a custom image which doesn't need the metadata service... This is in the post devices phase of bootup, so aborting means aborting the entire boot.

Logging goes to the system's console:

```
Welcome to the new instances experience!
We’re redesigning the EC2 console to make it easier to use. To switch between the old console and the new console, use the New EC2 Experience toggle above the navigation panel. We’ll release updates continuously based on customer feedback.

    EC2

Instances
i-0634d1a4114bb4140

    Get system log

Get system log Info
Review system log for instance i-0634d1a4114bb4140 as of Wed Nov 18 2020 16:35:47 GMT-0500 (EST)

[    0.147639] Cannot get hvm parameter CONSOLE_EVTCHN (18): -22!

[    0.666036] Cannot get hvm parameter CONSOLE_EVTCHN (18): -22!



[1;32m<<< NixOS Stage 1 >>>[0m



loading module xen-blkfront...

loading module xen-netfront...

loading module fuse...

loading module dm_mod...

loading module af_packet...

running udev...

Starting version 246

bringing up network interface eth0...

acquiring IP address via DHCP on eth0...

udhcpc: started, v1.31.1

udhcpc: sending discover

udhcpc: sending select for 172.31.38.31

udhcpc: lease of 172.31.38.31 obtained, lease time 3600

kbd_mode: KDSKBMODE: Inappropriate ioctl for device

%Gstarting device mapper and LVM...

CHANGED: partition=2 start=4096 old: size=4188160 end=4192256 new: size=6287327 end=6291423

waiting for device /dev/disk/by-label/nixos to appear...  WARNING: Device /dev/xvda2 not initialized in udev database even after waiting 10000000 microseconds.

  WARNING: Device /dev/xvda2 not initialized in udev database even after waiting 10000000 microseconds.

.

checking /dev/disk/by-label/nixos...

fsck (busybox 1.31.1)

[fsck.ext4 (1) -- /mnt-root/] fsck.ext4 -a /dev/disk/by-label/nixos

nixos: clean, 88235/131072 files, 370571/523520 blocks

resizing /dev/disk/by-label/nixos...

nixos: 88235/131072 files (0.1% non-contiguous), 370571/523520 blocks

resize2fs 1.45.5 (07-Jan-2020)

Resizing the filesystem on /dev/disk/by-label/nixos to 785915 (4k) blocks.

The filesystem on /dev/disk/by-label/nixos is now 785915 (4k) blocks long.



mounting /dev/disk/by-label/nixos on /...

(attempt 1/3) getting an EC2 instance metadata service v2 token...

*   Trying 169.254.169.253:80...

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                                 Dload  Upload   Total   Spent    Left  Speed


  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connection timed out after 1001 milliseconds


  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0

* Closing connection 0

curl: (28) Connection timed out after 1001 milliseconds

Warning: Problem : timeout. Will retry in 1 seconds. 3 retries left.

* Hostname 169.254.169.253 was found in DNS cache
```

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:8]: Good question. I'm not sure that should be part of this pr, but I added a commit doing so.

Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:8]: I pushed the commit but github is behind:

```
[grahamc@Petunia:~/projects/github.com/NixOS/nixpkgs/nixos]$ git push grahamc ec2-metadata-imdsv2
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 8 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 1.12 KiB | 383.00 KiB/s, done.
Total 5 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To github.com:grahamc/nixpkgs.git
   b1c1087e937..e8551fdd14f  ec2-metadata-imdsv2 -> ec2-metadata-imdsv2


[grahamc@Petunia:~/projects/github.com/NixOS/nixpkgs/nixos]$ git diff b1c1087e937..e8551fdd14f
diff --git a/nixos/release-combined.nix b/nixos/release-combined.nix
index d8b9a5f9b4b..1ef357463a1 100644
--- a/nixos/release-combined.nix
+++ b/nixos/release-combined.nix
@@ -49,6 +49,7 @@ in rec {
         [ "nixos.channel" ]
         (onFullSupported "nixos.dummy")
         (onAllSupported "nixos.iso_minimal")
+        (onAllSupported "nixos.amazonImage")
         (onSystems ["x86_64-linux"] "nixos.iso_plasma5")
         (onSystems ["x86_64-linux"] "nixos.iso_gnome")
         (onFullSupported "nixos.manual")
diff --git a/nixos/release-small.nix b/nixos/release-small.nix
index 6da2c59cedd..84bc71752c5 100644
--- a/nixos/release-small.nix
+++ b/nixos/release-small.nix
@@ -92,6 +92,7 @@ in rec {
       [ "nixos.channel"
         "nixos.dummy.x86_64-linux"
         "nixos.iso_minimal.x86_64-linux"
+        "nixos.amazonImage.x86_64-linux"
         "nixos.manual.x86_64-linux"
         "nixos.tests.boot.biosCdrom.x86_64-linux"
         "nixos.tests.containers-imperative.x86_64-linux"
```

**Comment by worldofpeace** [comment]
Oh, I see my review was requested. Is the goal to have this in the stable distro?
Anyways, I think graham is cool etc. and he do cool thing. that's my disposition on this pull request. Strongly in favor to put smiley faces, hearts, and sparkles.

**Comment by grahamc** [comment]
Yeah, I want to backport it so NixOS on AWS users can improve the security of their instances.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (andir) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:21]: Is there a chance to *not* have v2 enabled where this will always fail on some AWS account? If that is the case we might want to check what the error values there are.

If the metadata service wasn't available just yet would it be worse to delay the boot (and be useless?) or to delay it by 30s and have it potentially recover?

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (andir) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:38]: What is the size of that image? I think in the past we had concerns about large binary blobs being produced by hydra.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (andir) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:8]: I see. If I read the curl docs right it will double the delay, 1s, 2s, 4s, 8s, 16s.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:21]: > Is there a chance to not have v2 enabled where this will always fail on some AWS account? If that is the case we might want to check what the error values there are.

No. You can only have V1 & V2 _or_ V2 only: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html#instance-metadata-v2-how-it-works

> If the metadata service wasn't available just yet would it be worse to delay the boot (and be useless?) or to delay it by 30s and have it potentially recover?

The instance metadata service is never not available, and is provided by the hypervisor. The race here is "did we setup the network yet?" In general, I think this almost-never fails past the initial request, and so giving it up to 9 seconds seems generous already.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:8]: 1.4G. It is already built on every evaluation, this just makes it a blocker. This change grew the initrd by about 18M.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:21]: Instead of keeping the existing connect-timeout of 30s (woof) I'd rather extend the retry loop to be like 30 tries or something. But also I feel comfortable that the current configuration is good (or I wouldn't have PR'd it :sweat_smile: )

**Comment by grahamc** [comment]
Got a +1 from @andir on IRC.

**Review by AmineChikhaoui — APPROVED** [review]

**Comment by AmineChikhaoui** [comment]
FYI @endgame and @jonringer , this implements https://github.com/NixOS/nixpkgs/pull/96593 a bit differently I think.

**Comment by grahamc** [comment]
I provided Amine a copy of the AMI to test with NixOps, so let's merge once that is ready.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [inline nixos/modules/virtualisation/ec2-metadata-fetcher.nix:56]: Amine makes a very compelling point that the `1.0` here is a mistake :P. I'll get a new patch and AMI for him to test.

Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:56]: Update: on the contrary, these paths  still continue to work. Researching with Amine.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (grahamc) [inline nixos/modules/virtualisation/ec2-metadata-fetcher.nix:60]: I added a preflight check of the IMDS token, as apparently sometimes the instance metadata service takes a second to figure itself out.

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (andir) [inline nixos/modules/virtualisation/ec2-metadata-fetcher.nix:32]: At this point we might consider setting `PATH=${curl}/bin:$PATH` or something. This will safe use ~100 bytes or so ;-)

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (andir) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:60]: I hope we do not need more of these fallbacks.. It will soon escape the scope of where a shell script appears to be the right thing to me.

The commit message indicates it can take minutes for the metadata service to respond properly. Yet we are only waiting for to 30s for a response. Is that a trade-off between slow boot / unusable machines?

I wonder if the metadata service might also fail to hand out the "current" token during startup. Since we are already retrieving the instance-id should be validate something?

Comment (andir) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:60]: Ok, we talked on IRC and it is moment not minutes. I guess 30s are fine then?!

Comment (grahamc) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:60]: "minutes" was a mistake on my part. I updated it to say "moments". I too hope we don't need more, it is already growing to the limits of comfort. I think once we get evidence of a working token, we're good to go, but I'm happy to be shown wrong.

**Review by AmineChikhaoui — APPROVED** [review]
LGTM! thanks @grahamc

**Review by endgame — CHANGES_REQUESTED** [review]
I appreciate the ping. There are a couple of small things that I think need to change. Fixing IMDSv2 on nixos was actually a yakshave; I need the public IPv4 address from IMDS for something else I'm working on, and tried to fix IMDS while I was there.

Also, the metadata dir should probably cleared on each boot, because otherwise it can become outdated w.r.t actions people can take in the management console.

I think there's enough difference in #96593 that it's probably worth @grahamc having a look and seeing if there's anything he wants to lift: I constructed the PUT request using busybox `nc`, so I don't pull additional packages into initramfs, and the fetcher is parameterised so the EC2/openstack common parts have been DRYed up. Some of the comments might also be worth pulling across. I don't have a preflight check on the IMDS, though.

Sidenote: I must confess that I'm a little frustrated by what feels like a process failure here: I'd been trying to get help testing my PR for months, and there's so much to do on NixOS that duplicating work isn't great. But at the end of the day, progress is excellent and I'm glad to see this about to land :+1: :tada:

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (endgame) [inline nixos/modules/virtualisation/ec2-metadata-fetcher.nix:4]: This might not be "reasonable", but you can construct a PUT with busybox `nc`:
```
echo -ne 'PUT /latest/api/token HTTP/1.1\r\nHost: 169.254.169.254\r\nX-aws-ec2-metadata-token-ttl-seconds: 60\r\n\r\n' | nc 169.254.169.254 80 | tail -n 1
```

File: nixos/modules/virtualisation/ec2-metadata-fetcher.nix
Comment (endgame) [reply nixos/modules/virtualisation/ec2-metadata-fetcher.nix:60]: @andir I'm not sure there's a "current" token:

https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/ under "what's new in IMDSv2":

> The token is never stored by IMDSv2 and can never be retrieved by subsequent calls, so a session and its token are effectively destroyed when the process using the token terminates. There’s no limit on the number of requests within a single session, and there’s no limit on the number of IMDSv2 sessions.

So guess it generates some cryptographic doohickey each time you PUT?

**Comment by grahamc** [comment]
> I need the public IPv4 address from IMDS

I wonder why it needs to be part of the very early boot process, and not fetched as part of a service pre-start or something? We could add it, but I fear growing what is already a fairly gnarly script in a gnarly and unfortunate location.

> Also, the metadata dir should probably cleared on each boot, because otherwise it can become outdated w.r.t actions people can take in the management console.

This sounds like a good idea, but I'm a bit unsure about backporting to stable. It would be a pretty big change in behavior, potentially breaking workflows. Maybe an additional PR which only hits master, plus release notes. What do you think?

> I constructed the PUT request using busybox nc, so I don't pull additional packages into initramfs

I think the HTTP protocol is complicated enough to deserve using curl, and that implementing it with `nc` to save a few bytes is just going to cause pain for users / admins later on, and that it is well worth the extra ~20M it adds to the initrd. Even if it isn't so preferable to be doing this in the first place.

> the fetcher is parameterised so the EC2/openstack common parts have been DRYed up

Since they are different APIs with different implementations, I don't like the idea of trying to make it "dry" -- they're doing different things with APIs that look similar but evolve differently. I thought it made sense to split out the two, creating a less dynamic script, as opposed to a highly parameterized script in stage-1. I'd rather a bit of duplication here, as opposed to risking subtle bugs from dryness.

> Sidenote: I must confess that I'm a little frustrated by what feels like a process failure here: I'd been trying to get help testing my PR for months, and there's so much to do on NixOS that duplicating work isn't great. But at the end of the day, progress is excellent and I'm glad to see this about to land.

Yes. I agree with you.

**Comment by endgame** [comment]
Thanks for the quick response @grahamc .

> IPv4 from IMDS - does it need to happen here?

I have a service that needs to know its public IP address, and was setting it up as a NixOS module. So I want that file present on disk so that I can refer to it before the image boot script does `nixos-rebuild switch` as part of normal boot. If there is a better place to achieve what I want?

> Clear metadata on boot - is this actually a scary change?

Agree with your suggestion - this should happen later and go into release notes, so we don't blow up someone's cloud deployment by accident. I don't think that's that likely, but likelihood × severity = risk, and it's high enough that it's worth being cautious. I can do a separate PR for this.

> DRYing openstack/ec2 metadata fetchers

Openstack explicitly advertises their IMDS as [EC2-compatible](https://docs.openstack.org/nova/ussuri/user/metadata.html#metadata-ec2-format). Though I'm not sure whether they have a token service on their IMDS, and maybe we say that means they've failed in that goal? In any case, maybe add a comment to keep the two fetchers in "reasonable" sync?

**Comment by grahamc** [comment]
re the IP address: would it be possible to define your service like this? :

```
systemd.services.foobar.script = ''
  ip=$(get-from-imds public_ipv4)
  exec my-program "$ip"
'';
```

If you can send that PR shortly after this one, we can merge it soon while we're all on top of it. BTW I've started #nixos-cloud on Freenode if you're interested in that sort of thing.

w.r.t. Openstack, I've found it really hit or miss if a deployment has an up to date set of components, or has even configured and deployed all the components. The variability there leaves me grayer than I was before ... :grimacing:

**Comment by grahamc** [comment]
Pending the discussion here, a backport PR: https://github.com/NixOS/nixpkgs/pull/104302

**Comment by endgame** [comment]
> re the IP address: would it be possible to define your service [to fetch the ip in a script]?

Probably; I hadn't thought about doing it that way. It's a low-pressure side project so I'll give it a go at some point. Definitely should not hold up this PR.

I'll look at clearing out the metadata directory on boot this weekend, hopefully.

That's all my concerns addressed, I think. Should we ship this, then?

**Comment by grahamc** [comment]
Thanks, @endgame! Let's go!

**Pull request merged**
